src/ReferenceList.php

```diff
@@ -59,31 +62,23 @@ public function __construct( $references = array() ) {
 	 * @throws InvalidArgumentException
 	 */
 	public function addReference( Reference $reference, $index = null ) {
-		if ( !is_int( $index ) && $index !== null ) {
-			throw new InvalidArgumentException( '$index must be an integer or null' );
+		if ( $index !== null && ( !is_int( $index ) || $index < 0 ) ) {
+			throw new InvalidArgumentException( '$index must be a non-negative integer or null' );
+		}
+
+		if ( $reference->isEmpty() ) {
+			return;
 		}
 
-		if ( $index === null || $index >= count( $this ) ) {
+		if ( $index === null || $index >= count( $this->references ) ) {
 			// Append object to the end of the reference list.
-			$this->attach( $reference );
+			$this->references[] = $reference;
 		} else {
 			$this->insertReferenceAtIndex( $reference, $index );
 		}
 	}
 
 	/**
-	 * @see SplObjectStorage::attach
-	 *
-	 * @param Reference $reference
-	 * @param mixed $data Unused in the ReferenceList class.
-	 */
-	public function attach( $reference, $data = null ) {
```

Comment:
I don't think it makes sense to list all the methods `SplObjectStorage` contains (http://php.net/manual/en/class.splobjectstorage.php). Isn't it enough to note that `ReferenceList` no longer extends that class?